amazon/ion/writer_binary_raw.py

```diff
@@ -0,0 +1,365 @@
+
+_LENGTH_FLOAT_64 = 0x08
+_LENGTH_FIELD_THRESHOLD = 14
+_LENGTH_FIELD_INDICATOR = 0x0E
+_NULL_INDICATOR = 0x0F
+
+_NULLS = [
+    _TypeIds.null | _NULL_INDICATOR,
+    _TypeIds.bool_false | _NULL_INDICATOR,
+    _TypeIds.pos_int | _NULL_INDICATOR,
+    _TypeIds.float | _NULL_INDICATOR,
+    _TypeIds.decimal | _NULL_INDICATOR,
+    _TypeIds.timestamp | _NULL_INDICATOR,
+    _TypeIds.symbol | _NULL_INDICATOR,
+    _TypeIds.string | _NULL_INDICATOR,
+    _TypeIds.clob | _NULL_INDICATOR,
+    _TypeIds.blob | _NULL_INDICATOR,
+    _TypeIds.list | _NULL_INDICATOR,
+    _TypeIds.sexp | _NULL_INDICATOR,
+    _TypeIds.struct | _NULL_INDICATOR
+]
+
+
+def _serialize_null(buf, ion_event):
+    buf.append(_NULLS[ion_event.ion_type])
+
+
+def _serialize_bool(buf, ion_event):
+    if ion_event.value:
+        buf.append(_TypeIds.bool_true)
+    else:
+        buf.append(_TypeIds.bool_false)
+
+
+def _write_length(buf, length, tid):
+    if length < _LENGTH_FIELD_THRESHOLD:
+        buf.append(tid | length)
+    else:
+        buf.append(tid | _LENGTH_FIELD_INDICATOR)
+        _write_varuint(buf, length)
+
+
+def _write_int_value(buf, tid, value):
+    value_buf = bytearray()
+    length = _write_uint(value_buf, value)
+    _write_length(buf, length, tid)
+    buf.extend(value_buf)
+
+
+# TODO This can be consolidated with writer_text's _serialize_scalar_from_string_representation_factory validation.
+def _validate_scalar_value(value, expected_type):
+    if not isinstance(value, expected_type):
+        raise TypeError('Expected type %s, found %s.' % (expected_type, type(value)))
+
+
+def _serialize_int(buf, ion_event):
+    value = ion_event.value
+    _validate_scalar_value(value, six.integer_types)
+    if value == 0:
+        buf.append(_Zeros.int)
+        return
+    if value < 0:
+        value = -value
+        tid = _TypeIds.neg_int
+    else:
+        tid = _TypeIds.pos_int
+    _write_int_value(buf, tid, value)
+
+
+def _serialize_float(buf, ion_event):
+    float_value = ion_event.value
+    _validate_scalar_value(float_value, float)
+    # TODO Assess whether abbreviated encoding of zero is beneficial; it's allowed by spec.
+    if float_value.is_integer() and float_value == 0:
+        buf.append(_Zeros.float)
+        return
+    # TODO Add an option for 32-bit representation (length=4) per the spec.
+    buf.append(_TypeIds.float | _LENGTH_FLOAT_64)
+    encoded = struct.pack('>d', float_value)  # '>' specifies big-endian encoding.
+    buf.extend(encoded)
+
+
+def _write_decimal_value(buf, exponent, coefficient, sign=0):
+    length = _write_varint(buf, exponent)
+    if coefficient:  # The coefficient is non-zero, so the coefficient field is required.
+        length += _write_int(buf, coefficient)
+    elif sign:  # The coefficient is negative zero.
+        buf.append(_INT_NEG_ZERO)
+        length += 1
+    # Else the coefficient is positive zero and the field is omitted.
+    return length
+
+
+def digits_to_coefficient(digits):
```

Comment:
I was thinking here, does it make sense to scale the decimal and convert it to an integer instead of doing the base-10 conversion yourself?

E.g.

```python
d = Decimal(...)
sign, digits, exp = d.as_tuple()
coefficient = d.scaleb(-exp).to_integral_value()
```